### ISSUE TITLE
Router and navigation

### ISSUE THREAD

**Issue by hdeshev** [NativeScript/nativescript-angular#18]
Figure out if we can integrate with the [Angular Router](https://github.com/angular/angular/tree/master/modules/angular2/router).

If the above isn't possible/feasible, offer a similar API.

**Comment by tjvantoll** [comment]
IMO integration with the router, or some alternative API, is an absolute necessity. I ran into this limitation about 30 minutes into trying to build a serious app, and others will as well.

**Comment by MiroValchev** [comment]
@hdeshev is there any work-around? I am new to Angular and maybe its me, but can't think of a way to have two pages and navigate between them without router. This won't do topmost().navigate("views/calendar") :)

We have a native app in the stores and want to re-write it from scratch with {N} + Angular 2. Not really complicated but still will have more than a single view.

**Comment by hdeshev** [comment]
We're working on [something](https://github.com/NativeScript/qsf-angular-nativescript/blob/hdeshev/page-route/app/app-page.ts#L29-L161) that seems promising, but has some issues on iOS. I'll probably document it, push it to master this week and improve it later.

**Comment by NathanWalker** [comment]
Very exciting! :)

**Comment by tjvantoll** [comment]
@MiroValchev I have [a workaround in my Groceries sample](https://github.com/NativeScript/sample-Groceries/blob/710de30fdfe8640cabb489fb64ac02c1af894926/app/views/login/login.ts#L46) as well, but it’s definitely a bit hacky for now, so I’m also anxiously awaiting @hdeshev’s magic too 😀

**Comment by VicMota** [comment]
Totally +1. Any progress on this?
I personally really like the $stateProvider way, which is how Angular ui-router (Angular 1.x) does the routing.

**Comment by matt4446** [comment]
There was a merge today into mater related to the router: 

https://github.com/NativeScript/nativescript-angular/commit/f0768cb397a829a8728c2109c272b671aceda4ca

nsRouterLink directive looks interesting. It should be all there to set up the route config + base now.

**Comment by anhoev** [comment]
When I change page, angular destroys all components , and renders them again .
Is it possible to change page without destroying the components .

**Comment by vakrilov** [comment]
@anhoev Can you share some more info. Are you using the `page-router-outlet`?
BTW, there is already a working sample [here](https://github.com/NativeScript/qsf-angular-nativescript/blob/master/app/app-page.ts#L20) for reference on how to use the page navigation.

**Comment by anhoev** [comment]
@vakrilov I am using page-router-outlet . I 've tried the qsf example but that one has the problem too.
I wrote two logging line in ExamplePreviewComponent 's constructor :
       ```
 console.log('Example Preview Component');
 console.log('Rerender here');
```

and every time , when i click back and click the example again, it will show two logs in console.
The root page seems to work ok , only child-page.
If you write only one time console.log, it will show the log after 6-7 times.
Angular 2 work very fast , man doesn't notice it but some components need to receive data from sever.

**Comment by matt4446** [comment]
I have encountered the same issue as @anhoev on the page double loading components after the navigation changes (in my own project) and have put some workarounds for some observable subscriptions to understand that the controls are being destroyed... I could just dispose of it.... but laziness :) (and it doesn't stop the double HTTP requests to get data). 

I have some odd styling artifacts after navigation as well. Stack Panels in one of my components will get some extra padding (only after the first navigation, when the action bar jumps in). Navigating somewhere and back again fixes it to look like normal again.

**Comment by vakrilov** [comment]
I think some more clarification about the navigation in NativeScript is needed.
When navigating to a new page, the old page is put in the navigation stack. Under the hood, the navigation uses the corresponding native APIs: `UINavigationController`(IOS) and `FragmentManager`(Android). So, the old page still lives in the navigation stack. That's why, when you navigate back to it, no constructors for the NG components are called - only the `activate` method.
On the other hand, when navigating back, the current page is not saved anywhere - it is discarded. That's why when you navigate forward to the same route a new host page is created and also a new instance of the component.

That should explain why the constructor for your start component is called only once , but when navigating forward and back to a second page - it is created every time. 

One thing you can do use the DI and a service to cache your data to avoid unneeded HTTP requests.

**Comment by vakrilov** [comment]
@matt4446 About that styling issue - it looks like a bug, Can you share some code so that we cane reproduce it.

**Comment by matt4446** [comment]
@vakrilov If I can. 
I changed my CSS rules (and containers) about a week ago so that it didn't reproduce on navigation, but I think that I remember what caused it (it's in github's history as well). Stack Panels seemed to inconsistently choose margin / padding based on if the nav bar was present on the page. That is a little presumptuous, however... I'll see if I can create it again.

**Comment by vakrilov** [comment]
@matt4446 Can you open another issue if you manage to reproduce it? Also we made some changes in the Page layout when showing/hiding the action bar so it is possible that this issue is fixed with the newer version of the NativeScript modules.

Closing the issue - there is an example with nested navigation [here](https://github.com/NativeScript/nativescript-angular/tree/master/ng-sample/app/examples/navigation) and we will start working on the docs soon.
